File: Tease AI/Services/ImageAccessor.vb
Option Strict On
Option Infer Off
Imports System.IO
Imports Tease_AI
Imports Tease_AI.My
Imports TeaseAI.Common
Imports TeaseAI.Common.Constants
Imports TeaseAI.Common.Data
Imports TeaseAI.Common.Interfaces.Accessors

Public Class ImageAccessor
    Implements IImageAccessor

    Public Sub New(configurationAccessor As IConfigurationAccessor)
        myConfigurationAccessor = configurationAccessor
    End Sub

    Public Function GetImageMetaDataList(source As ImageSource?, genre As ImageGenre?) As Result(Of List(Of ImageMetaData)) Implements IImageAccessor.GetImageMetaDataList
        Dim returnValue As List(Of ImageMetaData) = New List(Of ImageMetaData)()

        Dim containers As List(Of ImageContainer) = myConfigurationAccessor.GetApplicationConfiguration().ImageContainers

        For Each container As ImageContainer In containers
            Dim searchLevel As FileIO.SearchOption = If(container.UseSubFolders, FileIO.SearchOption.SearchAllSubDirectories, FileIO.SearchOption.SearchTopLevelOnly)

            If (genre.GetValueOrDefault(container.Genre) = container.Genre) AndAlso source.GetValueOrDefault(container.Source) = container.Source Then
                If (container.Source = ImageSource.Local) Then
                    If (container.Genre = ImageGenre.Liked) OrElse (container.Genre = ImageGenre.Disliked) Then
                        For Each foundFile As String In File.ReadAllLines(container.Path)
                            If (Not foundFile.ToLower().StartsWith("http")) Then
                                returnValue.Add(New ImageMetaData() With
                                            {
                                            .Genre = container.Genre,
                                            .ItemName = foundFile,
                                            .Source = container.Source
                                            })
                            End If
                        Next
                    Else
                        For Each foundFile As String In My.Computer.FileSystem.GetFiles(container.Path, searchLevel, "*.jpg")
                            returnValue.Add(New ImageMetaData() With
                            {
                            .Genre = container.Genre,
                            .ItemName = foundFile,
                            .Source = container.Source
                            })
                        Next
                    End If
                ElseIf container.Source = ImageSource.Remote Then
                    If (container.Genre = ImageGenre.Liked) OrElse (container.Genre = ImageGenre.Disliked) Then
                        For Each foundFile As String In File.ReadAllLines(container.Path)
                            If (foundFile.ToLower().StartsWith("http")) Then
                                returnValue.Add(New ImageMetaData() With
                                            {
                                            .Genre = container.Genre,
                                            .ItemName = foundFile,
                                            .Source = container.Source
                                            })
                            End If
                        Next
                    ElseIf (container.Genre = ImageGenre.Blog) Then
                        For Each blogFile As String In GetAvailableBlogFiles()
                            For Each foundFile As String In File.ReadAllLines(blogFile)
                                returnValue.Add(New ImageMetaData() With
                                            {
                                            .Genre = ImageGenre.Blog,
                                            .ItemName = foundFile,
                                            .Source = container.Source
                                            })
                            Next
                        Next
                    Else
                        Dim fileName As String = myPathUrlFileDir + container.Path
                        For Each foundFile As String In File.ReadAllLines(fileName)
                            returnValue.Add(New ImageMetaData() With
                                            {
                                            .Genre = container.Genre,
                                            .ItemName = foundFile,
                                            .Source = container.Source
                                            })
                        Next
                    End If
                End If
            End If
        Next

        Return Result.Ok(returnValue)
    End Function

    Private Function GetAvailableBlogFiles() As List(Of String)
        Dim checkList As String = myPathUrlFileDir + "..\URLFileCheckList.cld"

        Dim returnValue As List(Of String) = New List(Of String)()
        Using fs As New FileStream(checkList, FileMode.Open), binRead As New BinaryReader(fs)
            Do While fs.Position < fs.Length
                Dim fileName As String = binRead.ReadString()
                Dim enabled As Boolean = binRead.ReadBoolean()
                Dim fullFilePath As String = myPathUrlFileDir + fileName + ".txt"

                If File.Exists(fullFilePath) AndAlso enabled Then
                    returnValue.Add(fullFilePath)
                End If
            Loop
        End Using
        Return returnValue.Distinct().ToList()
    End Function

    Public Function SaveImageMetaData(imageMetaDatas As List(Of ImageMetaData)) As Result Implements IImageAccessor.SaveImageMetaData
        Throw New NotImplementedException()
    End Function

    Private mySystemImageDIr As String = System.Windows.Forms.Application.StartupPath + "\Images\System\"
    Private myPathUrlFileDir As String = mySystemImageDIr + "URL Files\"
    Private ReadOnly myConfigurationAccessor As IConfigurationAccessor
End Class

File: TeaseAI.Services/Accessors/ImageAccessor.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using TeaseAI.Common;
using TeaseAI.Common.Constants;
using TeaseAI.Common.Data;
using TeaseAI.Common.Interfaces;
using TeaseAI.Common.Interfaces.Accessors;
using TeaseAI.Data.Interfaces;

namespace TeaseAI.Services.Accessors
{
    public class ImageAccessor : IImageAccessor
    {
        public ImageAccessor(IConfigurationAccessor configurationAccessor
            , IPathsAccessor pathsAccessor
            , IImageMetaDataRepository imageMetaDataRepository
            , IMediaContainerService mediaContainerService)
        {
            _configurationAccessor = configurationAccessor;
            _pathsAccessor = pathsAccessor;
            _imageMetaDataRepository = imageMetaDataRepository;
            _mediaContainerService = mediaContainerService;
        }

        public List<ImageMetaData> Get(ImageSource? source, ImageGenre? genre)
        {
            return _imageMetaDataRepository.Get(source, genre);
        }

        public Result Update(IEnumerable<ImageMetaData> imageMetaDatas)
        {
            return _imageMetaDataRepository.Update(imageMetaDatas);
        }

        public void Initialize()
        {
            var applicationConfig = _configurationAccessor.GetApplicationConfiguration();
            var mediaContainers = _mediaContainerService.Get();
            var imageContainers = mediaContainers.Where(mc => mc.MediaTypeId == 1 && mc.SourceId == ImageSource.Local && mc.IsEnabled).ToList();
            foreach (var imageContainer in imageContainers)
            {
                var searchLevel = imageContainer.UseSubFolders ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly;

                var files = Directory.GetFiles(imageContainer.Path, "*.*", searchLevel).ToList();
                foreach (var file in files)
                {
                    if (File.Exists(file))
                    {
                        _imageMetaDataRepository.Create(
                            new ImageMetaData
                            {
                                MediaContainerId = imageContainer.Id,
                                ItemName = Path.GetFileNameWithoutExtension(file),
                                FullFileName = file,
                                SourceId = imageContainer.SourceId,
                                GenreId = imageContainer.GenreId,
                            });
                    }
                }
            }
        }

        private IEnumerable<string> GetAvailableBlogFiles()
        {
            var checkListFile = _pathsAccessor.GetSystemImages()
                + Path.DirectorySeparatorChar + "URLFileCheckList.cld";
            var urlFileDir = _pathsAccessor.GetSystemImages()
                + Path.DirectorySeparatorChar + "URL Files";

            var returnValue = new List<string>();
            using (var fs = new FileStream(checkListFile, FileMode.Open))
            using (var binRead = new BinaryReader(fs))
            {
                while (fs.Position < fs.Length)
                {
                    var fileName = binRead.ReadString();
                    var isEnabled = binRead.ReadBoolean();
                    var fullFilePath = urlFileDir
                        + Path.DirectorySeparatorChar + fileName + ".txt";

                    if (File.Exists(fullFilePath) && isEnabled)
                        returnValue.Add(fullFilePath);
                }
                return returnValue.Distinct().ToList();
            }
        }

        private List<string> GetFiles(ImageContainer container, SearchOption searchLevel)
        {
            if (!container.IsEnabled)
                return new List<string>();
            var applicationConfig = _configurationAccessor.GetApplicationConfiguration();
            if (container.Genre == ImageGenre.Liked || container.Genre == ImageGenre.Disliked)
                return File.ReadAllLines(container.Path)
                    .Where(f => f.StartsWith("http") && container.Source == ImageSource.Remote)
                    .ToList();

            if (container.Source == ImageSource.Local)
            {
                var folderPath = container.Path;
                return Directory.GetFiles(folderPath, "*.*", searchLevel).ToList();
            }

            var urlFileDir = _pathsAccessor.GetSystemImages()
                + Path.DirectorySeparatorChar + "URL Files"
                + Path.DirectorySeparatorChar + container.Path;
            return File.ReadAllLines(urlFileDir).ToList();
        }

        public void Create(List<ImageMetaData> images)
        {
            _imageMetaDataRepository.Create(images);
        }

        public Result<List<ImageMetaData>> GetImagesInContainer(int containerId)
        {
            return _mediaContainerService.Get(containerId)
                .OnSuccess(mc =>
                {
                    var dbImages = _imageMetaDataRepository.Get(mc.SourceId, mc.GenreId);
                    var fileImages = GetFiles(mc);
                    foreach (var file in fileImages)
                    {
                        if (dbImages.All(imd => imd.FullFileName != file))
                        {
                            var newImage = new ImageMetaData
                            {
                                FullFileName = file,
                                ItemName = Path.GetFileNameWithoutExtension(file),
                                MediaContainerId = mc.Id,
                                GenreId = mc.GenreId,
                                SourceId = mc.SourceId,
                            };
                            _imageMetaDataRepository.Create(newImage)
                                .OnSuccess(imd =>
                                {
                                    dbImages.Add(imd);
                                });
                        }
                    }

                    return Result.Ok(dbImages.OrderBy(imd => imd.ItemName ).ToList());
                });

        }

        private List<string> GetFiles(MediaContainer mediaContainer)
        {
            if (!Directory.Exists(mediaContainer.Path))
                return new List<string>();
            //"*.png,*.jpg,*.gif,*.bmp,*.jpeg"
            var searchOption = mediaContainer.UseSubFolders ? SearchOption.AllDirectories : SearchOption.TopDirectoryOnly;
            var temp = Directory.GetFiles(mediaContainer.Path, "*.jpg", searchOption);
            return temp.ToList();
        }

        public List<ImageMetaData> GetImagesWithTag(ItemTagId itemTagId) => _imageMetaDataRepository.GetImagesWithTag(itemTagId);

        private readonly IConfigurationAccessor _configurationAccessor;
        private readonly IPathsAccessor _pathsAccessor;
        private readonly IImageMetaDataRepository _imageMetaDataRepository;
        private readonly IMediaContainerService _mediaContainerService;
    }
}
